lib/pos_mempool.go

```diff
@@ -0,0 +1,394 @@
+	}
+	// Reset the transaction register and the ledger.
+	dmp.txnRegister.Reset()
+	dmp.ledger.Reset()
+
+	dmp.emitEvents = false
+	dmp.status = PosMempoolStatusNotRunning
+}
+
+func (dmp *PosMempool) IsRunning() bool {
+	return dmp.status == PosMempoolStatusRunning
+}
+
+// ProcessMsgDeSoTxn validates a MsgDeSoTxn transaction and adds it to the mempool if it is valid.
+// If the mempool overflows as a result of adding the transaction, the mempool is pruned.
+func (dmp *PosMempool) ProcessMsgDeSoTxn(txn *MsgDeSoTxn) error {
+	if dmp.status == PosMempoolStatusNotRunning {
+		return errors.Wrapf(MempoolErrorNotRunning, "PosMempool.ProcessMsgDeSoTxn: ")
+	}
+
+	// First, validate that the transaction is properly formatted according to BalanceModel.
+	latestBlockHeight := dmp.latestBlockNode.Height
+	if err := ValidateDeSoTxnSanityBalanceModel(txn, uint64(latestBlockHeight), dmp.params, dmp.globalParams); err != nil {
+		return errors.Wrapf(err, "PosMempool.AddTransaction: Problem validating transaction sanity")
+	}
+
+	// Construct the MempoolTx from the MsgDeSoTxn.
+	mempoolTx, err := NewMempoolTx(txn, uint64(latestBlockHeight))
+	if err != nil {
+		return errors.Wrapf(err, "PosMempool.AddTransaction: Problem constructing MempoolTx")
+	}
+
+	// Check transaction signature
+	if _, err := dmp.latestBlockView.VerifySignature(txn, latestBlockHeight); err != nil {
+		return errors.Wrapf(err, "PosMempool.AddTransaction: Signature validation failed")
+	}
+
+	// Add the transaction to the mempool.
+	if err := dmp.AddTransaction(mempoolTx); err != nil {
+		return errors.Wrapf(err, "PosMempool.AddTransaction: Problem adding transaction to mempool")
+	}
+	return nil
+}
+
+// AddTransaction is the main function for adding a new transaction to the mempool.
+func (dmp *PosMempool) AddTransaction(txn *MempoolTx) error {
+	if dmp.status == PosMempoolStatusNotRunning {
+		return errors.Wrapf(MempoolErrorNotRunning, "PosMempool.AddTransaction: ")
+	}
+
+	dmp.Lock()
+	defer dmp.Unlock()
+
+	// Add the transaction to the mempool and then prune if needed.
+	if err := dmp.addTransactionNoLock(txn); err != nil {
+		return errors.Wrapf(err, "PosMempool.AddTransaction: Problem adding transaction to mempool")
+	}
+
+	if err := dmp.pruneNoLock(); err != nil {
+		glog.Errorf("PosMempool.AddTransaction: Problem pruning mempool: %v", err)
+	}
+
+	return nil
+}
+
+func (dmp *PosMempool) addTransactionNoLock(txn *MempoolTx) error {
+	userPk := NewPublicKey(txn.Tx.PublicKey)
+	txnFee := txn.Tx.TxnFeeNanos
+	latestBlockHeight := dmp.latestBlockNode.Height
+
+	// Validate that the user has enough balance to cover the transaction fees.
+	spendableBalanceNanos, err := dmp.latestBlockView.GetSpendableDeSoBalanceNanosForPublicKey(userPk.ToBytes(), latestBlockHeight)
+	if err != nil {
+		return errors.Wrapf(err, "CheckBalanceIncrease: Problem getting spendable balance")
+	}
+	if err := dmp.ledger.CanIncreaseBalance(*userPk, txnFee, spendableBalanceNanos); err != nil {
```

Comment:
Isn't this actually a balance DECREASE? Okay so I see what the ledger is doing now - can we rename these functions so it's more clear what they do. It hink something like `CanIncreaseTxnFeeBalance` or something of that nature.

Reply:
Hmm I think it should be an increase. We're trying to add a new transaction to the mempool, so we're checking if the user's reserved balance (in fees) + the new txn fee is within user's spendable balance.

Reply:
I think the name is a little confusing is all. @tholonious - how do you feel about the naming convention here? If it's clear to you, then I'm happy to keep it as is so we don't have a long and unwieldy name. I know what it's doing now, but the name didn't make it clear to me at first glance.